src/core/electrostatics_magnetostatics/elc.cpp

```diff
@@ -1460,7 +1460,7 @@ void ELC_P3M_restore_p3m_sums(ParticleRange const &particles) {
 
   MPI_Allreduce(node_sums, tot_sums, 3, MPI_DOUBLE, MPI_SUM, comm_cart);
 
-  p3m.sum_qpart = static_cast<int>(tot_sums[0] + 0.1);
+  p3m.sum_qpart = static_cast<int>(std::round(tot_sums[0]));
```

Comment:
sure about this? the ```+0.1``` probably was added to produce ```std::ceil```?

Reply:
that's precisely why I didn't touch the rounding operation in the first commit :smiley:

We can argue about the intent of the original developer, to me `std::round()` makes sense because in `(int) (x + 0.1)`, the `0.1` will be larger than machine precision for a double (for small x values, obviously), so the expression is effectively `X + 0.1` (with `X` the "real" value), and the static cast to int will lead to a narrowing conversion, so we get X. Since `sum_qpart` is supposed to contain the integer number of particles that contribute to the sum of charges, this checks out. Using `std::ceil` would make the code sensitive to machine precision, e.g. if the double stores X.0000000001 it will become X+1. This would be a silent bug, because the tests ignore the value of `sum_qpart` (incrementing its value changes nothing to the ELC tests outcome).

This `(int) (x + 0.1)` logic is quite fragile, but it's used in the P3M and DP3M code as well, so I'll revert this change to the original version.